117.md

```diff
@@ -0,0 +1,361 @@
+# NIP-117
+
+## The Double Ratchet Algorithm
+
+The Double Ratchet is a key rotation algorithm for secure private messaging.
+
+It allows us to 1) communicate on Nostr without revealing metadata (who you are communicating with and when), and 2) keep your message history and future messages safe even if your main Nostr key is compromised.
+
+Additionally, it enables disappearing messages that become undecryptable when past message decryption keys are discarded after use.
+
+See also: [NIP-118](./118.md): Nostr Double Ratchet Invites
+
+### Overview
+
+"Double ratchet" means we use 2 "ratchets": cryptographic functions that can be rotated forward, but not backward: current keys can be used to derive next keys, but not the other way around.
+
+Ratchet 1 uses Diffie-Hellman (DH) shared secrets and is rotated each time the other participant acknowledges a new key we have sent along with a previous message.
+
+Ratchet 2 generates encryption keys for each message. It rotates after every message, using the previous message's key as input (and the Ratchet 1 key when it rotates). This process ensures forward secrecy for consecutive messages from the same sender in between Ratchet 1 rotations.
+
+## Nostr implementation
+
+We implement the Double Ratchet Algorithm on Nostr similarly to Signal's [Double Ratchet with header encryption](https://signal.org/docs/specifications/doubleratchet/#double-ratchet-with-header-encryption), but encrypting the message headers with [NIP-44](./44.md) conversation keys instead of symmetric header keys.
+
+Ratchet 1 keys are standard Nostr keys. In addition to encryption, they are also used for publishing and subscribing to messages on Nostr. As they are rotated and not linked to public Nostr identities, metadata privacy is preserved.
+
+## Nostr event format
+
+### Message
+
+#### Outer event
+
+```typescript
+{
+  kind: 1060,
+  content: encryptedInnerEvent,
+  tags: [["header", encryptedHeader]],
+  pubkey: ratchetPublicKey,
+  created_at,
+  id,
+  sig
+}
+```
+
+We subscribe to Double Ratchet events based on author public keys which are ephemeral — not used for other purposes than the Double Ratchet session. We use the regular event kind `1060` to differentiate it from other DM kinds, retrieval of which may be restricted by relays.
+
+The encrypted header contains our next nostr public key, our previous sending chain length and the current message number.
+
+The events should be sent to the recipient's inbox relays.
```

Comment:
Shouldn't this use the DM relays? Otherwise, if I am the only user of my public inbox, it's possible to identify DMs sent to me there.